### ISSUE TITLE
Add a verification flag to ignore trailing dot in hostname 

### ISSUE THREAD

**Issue by brysonli12** [openssl/openssl#11560]
Can the team add a verification flag to openssl to handle the trailing dot (if it exists) in hostname appropriately (for the hostname check). This is needed because sometimes we need to have the trailing dot in the hostname for DNS resolution to work properly (and for security), but that would cause the certificate SAN fields to not match the hostname (since we have the trailing hostname).

So for example, let's say that I am contacting
"example.com." but my certificate has "example.com" on it, proposed change would match the certificate correctly with the hostname and allow for a connection.

**Comment by t8m** [comment]
I do not see a reason why the trailing dot should not be ignored unconditionally, i.e. no additional verification flag should be needed.

**Comment by brysonli12** [comment]
> When doing hostname verification against an X.509 certificate, a trailing dot (period) in the hostname is matched against the certificate.  But the trailing dot should only be applied to the DNS lookup not the certificate match.
> 
> Conversely, a certificate that has a trailing dot in its commonName (probably rare but allowed) should match a hostname without the trailing dot.  As the ssl library is written now, both cases fail.
> 
> Web browsers allow a trailing dot in the URI and will accept a certificate even when the certificate doesn't have a trailing dot.  For example, visit Google with trailing dot (https://www.google.com./) from a browser of your choice and check certificate.  It should check out as valid.  -from another post online related to the topic

Also, contacting a URL without a trailing dot can be ambiguous (ie. ai.university vs ai.unversity.edu). That is why we may want to contact a url using a trailing dot. But we do not typically put hostname with trailing dot in a certificate, we just put the hostname. Let me give you an example where this will be a problem.

In Python for example, if we contact a url with a trailing dot, the hostname verification directly using openssl API only uses that url with the trailing dot for verification (so we would compare "hostname.com." with "hostname.com" in the certificate which would not match.

Now, you may say that can just be fixed on the python side, but the python developers said they would not touch the ssl verification stuff (which is why they switched to directly calling the openssl API for this). Let me know your thoughts on this. How do you think this can practically be fixed? On OpenSSL? on Python? somewhere else?

**Comment by t8m** [comment]
Maybe you misunderstood my comment - I am basically saying this should be fixed on OpenSSL side and there is IMO no need to add a flag for that.

**Comment by t8m** [comment]
What I am not sure about is whether this can be considered simply a bug or it is rather a feature.

**Comment by abohosam1772** [comment]
Google

**Comment by brysonli12** [comment]
Ok, thanks @t8m I did misunderstand your comment

**Comment by brysonli12** [comment]
I think it would be considered a bug because expected behavior is to ignore trailing dot in hostname that you are comparing against the ones in the certificate but at the same time this is an important part of ssl verification (and so would be a feature).

**Comment by mattcaswell** [comment]
Perhaps @vdukhovni has an opinion on this?

**Comment by joekohlsdorf** [comment]
> To be clear, if OpenSSL is calling `SSL_set1_host()` because we're establishing the connection to the host via some form of `BIO_connect` type call that takes hostnames, and uses `SSL_set1_host()` and sets the SNI name internally, ... then we're responsible for stripping off any trailing `.` in the hostname before using it as an SNI name or passing it to `SSL_set1_host()` (but not when using it in `getaddrinfo(3)`).

This is the case for cpython, see code here: https://github.com/python/cpython/blob/d5bd32f/Modules/_ssl.c#L778 (main branch as of today)

We'd like to use FQDNs to avoid unnecessary DNS queries but with the current OpenSSL implementation we can't.

**Comment by vdukhovni** [comment]
Nothing in OpenSSL prevents the application from passing a dot-stripped name to `SSL_set1_host()` while separately using a trailing dot with `getaddrinfo(3)`.  If the question is about higher level BIO APIs that take a name and do all the connection management, the dot stripping prior to calling `SSL_set1_host()` can be done there.

There's no need to second guess intent at the lower layers of the library.

**Comment by joejulian** [comment]
> There's no need to second-guess intent at the lower layers of the library.

I disagree. I keep running in to issues where various applications have broken implementations because this is so ambiguous. It also doesn't help that the SNI standard expects it but the SAN doesn't.

Adding the trailing dot is not expected in the SAN per RFC 5280 4.2.1.6 paragraph 6:

```
When the subjectAltName extension contains a domain name system
label, the domain name MUST be stored in the dNSName (an IA5String).
The name MUST be in the "preferred name syntax", as specified by
Section 3.5 of [RFC1034] and as modified by Section 2.1 of
[RFC1123]. Note that while uppercase and lowercase letters are
allowed in domain names, no significance is attached to the case. In
addition, while the string " " is a legal domain name, subjectAltName
extensions with a dNSName of " " MUST NOT be used. Finally, the use
of the DNS representation for Internet mail addresses
(subscriber.example.com instead of [subscriber@example.com](mailto:subscriber@example.com)) MUST NOT
be used; such identities are to be encoded as rfc822Name. Rules for
encoding internationalized domain names are specified in Section 7.2.
```

an IA5String is defined in [RFC 1034 3.5](https://datatracker.ietf.org/doc/html/rfc1034#section-3.5) makes no mention of a trailing dot and the examples do not show one:
```
The following syntax will result in fewer problems with many
applications that use domain names (e.g., mail, TELNET).

<domain> ::= <subdomain> | " "

<subdomain> ::= <label> | <subdomain> "." <label>

<label> ::= <letter> [ [ <ldh-str> ] <let-dig> ]

<ldh-str> ::= <let-dig-hyp> | <let-dig-hyp> <ldh-str>

<let-dig-hyp> ::= <let-dig> | "-"

<let-dig> ::= <letter> | <digit>

<letter> ::= any one of the 52 alphabetic characters A through Z in
upper case and a through z in lower case

<digit> ::= any one of the ten digits 0 through 9
```
The verbiage of this section is very non-committal but I think it's fair for people to read this and expect that a SAN should not end in a dot.

**Comment by nhorman** [comment]
it seems like this still has some activity around it

@joekohlsdorf  @joejulian @brysonli12 

could one of you please add a comment with a proposed change to openssl so that it can be considered?  What I'm reading above sounds to me like the ask is that domain names for the purposes of certificate parsing/validation/etc have any trailing dots ignored, but I'd like to be sure

**Comment by joejulian** [comment]
You are correct. Taking dots should be ignored.

**Comment by gaelgatelement** [comment]
For the record, Golang 1.25.2 [tried to enable Strict validation of SANs in certificates](https://github.com/golang/go/issues/75828). They rolled back as it caused wide breakage, but they plan to enable it again later. 

Current OpenSSL behaviour is directly in conflict with this change. Golang would reject SANs with trailing dots, while the workaround to this OpenSSL bug is to issue certificates with trailing dots in SANs.

**Comment by vdukhovni** [comment]
> For the record, Golang 1.25.2 [tried to enable Strict validation of SANs in certificates](https://github.com/golang/go/issues/75828). They rolled back as it caused wide breakage, but they plan to enable it again later.

Yes, their original change was a mistake. It is fine to ignore a non-compliant SAN in the certificate, but other SANs should continue to be processed.

> Current OpenSSL behaviour is directly in conflict with this change. Golang would reject SANs with trailing dots, while the workaround to this OpenSSL bug is to issue certificates with trailing dots in SANs.

The right work-around is not issue SANs with trailing `.`s, but to fix the applications that are using names with trailing dots as SNI hostnames and/or as DNS-ID reference identifiers.  These applications should not do that.

But, as I said before, it makes reasonable sense for OpenSSL to strip the trailing dot in both cases, and pass the resulting "dot-canonicalised" form to the relevant underlying APIs (or do that when storing a copy of the provided the name for subsequent use).

What we SHOULD NOT do, is change the code that is performing the actual cert checks. It will continue to compare the stored reference identifier with the provided identifier without ad hoc heuristics.
